Sources/Codability/GenericDecodingContainer.swift

```diff
@@ -11,6 +27,26 @@ extension KeyedDecodingContainer {
         return try decodeIfPresent(T.self, forKey: key)
     }
 
+    /// Decode a heterogeneous list of objects for a given family.
+    /// - Parameters:
+    ///     - heterogeneousType: The decodable type of the list.
+    ///     - family: The ClassFamily enum for the type family.
+    ///     - key: The CodingKey to look up the list in the current container.
+    /// - Returns: The resulting list of heterogeneousType elements.
+    public func decode<T : Decodable, U : ClassFamily>(_ heterogeneousType: [T].Type, ofFamily family: U.Type, forKey key: K) throws -> [T] {
+        var container = try self.nestedUnkeyedContainer(forKey: key)
+        var list = [T]()
+        var tmpContainer = container
+        while !container.isAtEnd {
+            let typeContainer = try container.nestedContainer(keyedBy: Discriminator.self)
+            let family: U = try typeContainer.decode(U.self, forKey: U.discriminator)
+            if let type = family.getType() as? T.Type {
```

Comment:
This will always just decode `T` and not any subclasses. Decodable needs compile time types to know what to decode, not runtime types

Reply:
Did you had time to run UT ? we already using this approach and it works. T here will return the subtype for example Beer or Drink

Reply:
`getType` returns a compiler unknown type which you're casting to `T`. That means `tmpContainer.decode(type)` will always just be the superclass `T` (`Drink` in your tests)

Reply:
Yeah you are right, but it's no more the case under Swift4. The cast here is particularly important because the function `getType` returns the `AnyObject` type which is not decodable.
If you think otherwise, let me know, it may be worth thinking about it a little further and adding a suggestion for improvement.

Reply:
A think another approach would be required. Maybe some custom enum with associated types that adheres to a protocol

Reply:
that's what we have already, an enum that has variants for the parsable type (see unit test file) and that should adhere to our class ClassFamily protocol.

Reply:
Yes, but your enum doesn’t have any static types. I think if you run your tests you will see the problem

Reply:
I believe you mean the return type of the method getType in DrinkFamily enum, I guess you cannot have a static ( specific type ) as supposed to be as generic as possible.

Reply:
The decoding would have to be inside the enum where the specific types can be passed to the decoder 👍

Reply:
I've enabled the tests on CI for forks, so if you push again you can see the tests failing

Reply:
For sure it will fails, because this approach works only on Xcode 9.3.1 and above 
<img width="647" alt="unit-tests" src="https://user-images.githubusercontent.com/35116901/42756740-6d1b41d2-88f5-11e8-96fd-5150844485e8.png">

Reply:
I see that they are green 👍

Reply:
Oh cool you fixed it 👍  
I didn't think the decoder would decode a polymorphic type, but it seems to do the trick, nice! 😄

Reply:
Yeah, I forgot to put the right test case I Index.
So finally, you are convinced ?

Reply:
Hello again,
So what you've decided ?